Consider altChunk parts in Flat OPC transformation

ThomasBarnekow opened the pull request:
This commit enhances the

  ```XDocument OpenXmlPackage.ToFlatOpcDocument(XProcessingInstruction)```

method to consider "altChunk parts" (AlternativeFormatInputParts),
which must have their content rendered in binary form even if it is XML
content (e.g., XML, XHTML).

Fixes #525.

ThomasBarnekow commented:
@twsouthwick, how can I check why this failed? I don't have access to the Azure DevOps project. 

All unit tests passed on my local machine, so I don't know what might have happened.

ThomasBarnekow commented:
I might have found another "bug" in `Directory.Build.props`, which still lists `net35`.

```
<ProductTargetFrameworks>netstandard1.3;netstandard2.0;net35;net40;net46</ProductTargetFrameworks>
```

After setting `ProjectLoadStyle` to `All`, the build failed also on my machine. This was due to the fact that `net35` does not provide the `ISet<T>` interface.

ThomasBarnekow commented:
So I added a fix for net35 and the solution was built successfully and all unit tests passed. What goes wrong? I don't have access to the details.

rominator1983 commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -49,68 +54,85 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
-                    Package.GetParts().Select(part => GetContentsAsXml(part))));
+                    Pkg + "package",
+                    new XAttribute(XNamespace.Xmlns + "pkg", Pkg.ToString()),
+                    Package.GetParts().Select(p => GetContentsAsXml(p, altChunkPartUris))));
```
I personally do not like abbreviations in code but I don't know about the style guide of this project. So that's up to you...

ThomasBarnekow replied:
We might even be able to fit in `art` twice ;-)

ThomasBarnekow commented:
@rominator1983, do you have access to those details which might tell us why the check failed?

twsouthwick commented:
Good point about access. We're building it on the office dev ops instance so we can sign stuff... but we can't give access. I'll look into how we can expose that.

twsouthwick commented:
As for .NET 3.5, I do want to continue to support it, but it does limit some things. What I end up doing is adding stubs for the types to get things working. You can see under the `System` folder, a number of these, such as `IReadOnlyLIst` and others.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -49,68 +54,85 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
-                }
+                // The following expression creates the base64String, then chunks
+                // it to lines of 76 characters long.
+                string base64String = Convert.ToBase64String(byteArray)
```
Can you move this to a separate method?

ThomasBarnekow replied:
Done.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -270,5 +294,79 @@ protected static Package FromFlatOpcDocumentCore(XDocument document, Package pac
+        /// <summary>
+        /// Represents the collection of internal package relationships.
+        /// </summary>
+        private class InternalPackageRelationships : IEnumerable<PackageRelationship>
```
Could this be done as an iterator function instead?

ThomasBarnekow replied:
Yes, I will remove this general-purpose class and replace it with a much simpler, altChunk-specific expression.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -270,5 +294,79 @@ protected static Package FromFlatOpcDocumentCore(XDocument document, Package pac
+#else
+            private void Enumerate(PackageRelationship r, ISet<PackageRelationship> relationships)
+            {
+                if (!relationships.Add(r))
```
The logic here is different. How come you need `ISet<>`? Why not just ask for the HashSet<T> that is generated above? These are all internal, so no worry about leaking details

ThomasBarnekow replied:
I typically try to use the most abstract types to foster reuse. In this case, that would be `ISet<T>`. But that simply does not exist in `net35` (released 12 years ago).

twsouthwick replied:
Yes, I agree with trying to do that, but do you *need* it to be `ISet<T>` here? What's the limitation of `ICollection<T>`?

That's the challenge with trying to support `net35` :( We'll probably drop it at some point in the near future, once we can figure out usage and the impact that will have.

ThomasBarnekow replied:
No, it does not have to be `ISet<T>`. However, if it is not some kind of set, the `Add(T)` method will not return a flag telling me whether the element was added. If it is an `ICollection<T>`, I will have to check for containment first. However, if the underlying collection is a `HashSet<T>`, that will also be O(1). But I need two method calls rather than just one.

As I'm now getting rid of this class, there is no issue anymore. I borrowed it from another use case where net35 is no factor.

twsouthwick commented:
Thanks for taking a look at this. A few comments about some of the structure. I tend to avoid too much LINQ as recent profiling I've done has identified it as hot spots. I'm fine with starting off with it for getting correctness, but keeping them in separate methods will help during profiling to identify any specific queries that are problematic.

Also, I don't see any tests being added? Can you add some of those? Preferably more unit test style without needing a full file; just build up the components needed and the expected outcome.

ThomasBarnekow commented:
@twsouthwick, thanks for your feedback. 

Please have a look at the unit test in the `OpenXmlPackageTests` class in the `DocumentFormat.OpenXml.Packaging.Tests` project.

twsouthwick commented:
Must have missed the test - sorry about that. You can disregard that comment.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -270,5 +294,79 @@ protected static Package FromFlatOpcDocumentCore(XDocument document, Package pac
+                public int GetHashCode(PackageRelationship obj)
+                {
+                    string uriString = obj.SourceUri + ":" + obj.TargetUri;
+                    return uriString.GetHashCode();
```
I'm updating an internal method to support this. Once #660 is merged, you can build a hashcode with comparers.

ThomasBarnekow replied:
I'll get rid of the class, so that's no longer needed.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -14,10 +16,13 @@ namespace DocumentFormat.OpenXml.Packaging
     /// Represents a base class for strong typed Open XML document classes.
     /// </summary>
-    public abstract partial class OpenXmlPackage : OpenXmlPartContainer, IDisposable
+    public abstract partial class OpenXmlPackage
```
I just realized you're removing class hierarchy - that's a breaking change. Why are you doing this?

ThomasBarnekow replied:
I don't remove class hierarchy. In a **partial class**, we don't have to mention the base class and interfaces again. I could add those back in if you like.

twsouthwick replied:
Ah - did know it was defined on the other partial class. Yeah, I'm fine with not having it in this file (no need for duplication)

ThomasBarnekow commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -49,68 +54,85 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
+        /// <param name="altChunkPartUris">The collection of AlternativeFormatInputPart URIs.</param>
         /// <returns>The corresponding <see cref="XElement"/>.</returns>
-        private static XElement GetContentsAsXml(PackagePart part)
+        private static XElement GetContentsAsXml(PackagePart part, IEnumerable<Uri> altChunkPartUris)
```
The formal parameter does not have to be a `HashSet<T>` to utilize the O(1) for direct lookup. While it turns out that an `IEnumerable<T>` is not the optimal choice, because `Contains(T)` is an extension method rather than a method defined by the interface, the `Contains(T)` implementation of `HashSet<T>` will be used when it is passed as an actual parameter.

Having said that, I also think it would be better to use `HashSet<T>` or at a minimum `ICollection<T>`, which defines the `Contains(T)` method for `HashSet<T>`.

ThomasBarnekow commented:
Made the changes requested by @twsouthwick. Rebased on master and fixed up some commits.

ThomasBarnekow commented:
@twsouthwick, can you tell me what went wrong here? Setting `ProjectLoadStyle` to `All`, the solution is successfully built and all unit tests pass on my machine.

twsouthwick commented:
I have warnings as errors turned on. I recently added some FxCop rules to enforce StringComparer, so that's what's being raised:

```
src\DocumentFormat.OpenXml\Packaging\OpenXmlPackage.FlatOpc.cs(89,17): Error CA1307: The behavior of 'string.EndsWith(string)' could vary based on the current user's locale settings. Replace this call in 'DocumentFormat.OpenXml.Packaging.OpenXmlPackage.GetContentsAsXml(System.IO.Packaging.PackagePart, System.Collections.Generic.HashSet<System.Uri>)' with a call to 'string.EndsWith(string, System.StringComparison)'.
```

Not having access to the build is not good. I may create a separate devops instance for CI builds that I can make public. I'll look into that.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -49,71 +56,89 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
+        private static string ToChunkedBase64String(byte[] byteArray)
+        {
+            return Convert.ToBase64String(byteArray)
+                .Select((@char, index) => new { Character = @char, Chunk = index / 76 })
```
Can you make the `76` a const with a descriptive name as to what it is (ie `CharactersPerRow` or something?). I think it's something we'll forget what it is at some point :)

ThomasBarnekow replied:
Done. The 76 characters are based on the MIME specification, which defines a maximum line length of 76 characters for Base64-encoded lines.

twsouthwick replied:
Good to know - can you ensure there's a comment to that effect?

ThomasBarnekow replied:
That's already included, of course ;-)

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -49,71 +56,89 @@ protected XDocument ToFlatOpcDocument(XProcessingInstruction instruction)
             }
         }
 
+        private static string ToChunkedBase64String(byte[] byteArray)
```
I swear I've seen something like this in the code base before. Do you know if it's done else where? It may be useful if so to make it a general purpose utility

ThomasBarnekow replied:
I looked around but could not find anything else. You might have seen it in the Flat OPC code before. I turned it into a function, which I then beautified (e.g., formatted, renamed parameters).

twsouthwick replied:
sounds good.

twsouthwick commented on src/DocumentFormat.OpenXml/Packaging/OpenXmlPackage.FlatOpc.cs:
```diff
@@ -167,61 +192,63 @@ protected static Package FromFlatOpcDocumentCore(XDocument document, Package pac
-                            System.Convert.FromBase64CharArray(
-                                base64CharArray, 0, base64CharArray.Length);
+                        var base64StringInChunks = (string) xmlPart.Element(Pkg + "binaryData");
+                        char[] base64CharArray = base64StringInChunks.Where(c => c != '\r' && c != '\n').ToArray();
```
Can you move this out to a function as well?

ThomasBarnekow replied:
Done. And you got me started on a more complete refactoring of the other methods as well.

ThomasBarnekow commented:
@twsouthwick, is it possible that you turned those FxCop rules on after my previous build? That would explain it because I did not change that line of code with the warning.

twsouthwick commented:
It was sometime yesterday (Pacific time) in the afternoon. Highly possible

twsouthwick commented:
I opened #666 as a tracking item for seeing build failures

ThomasBarnekow commented:
@twsouthwick, unless you have further input, this would be it for now. I've updated CHANGELOG.md and also beautified the Flat OPC-related partial classes (where I also removed the redundant base type references.
